pkg/controller/volume/attachdetach/cache/desired_state_of_world.go

```diff
@@ -254,7 +254,14 @@ func (dsw *desiredStateOfWorld) AddPod(
 		}
 		dsw.nodesManaged[nodeName].volumesToAttach[volumeName] = volumeObj
 	}
-	if _, podExists := volumeObj.scheduledPods[podName]; !podExists {
+	// A Pod delete event followed by an immediate Pod add event may be merged
+	// into a Pod update event (from the shared informer). See
+	// https://github.com/kubernetes/client-go/blob/master/tools/cache/shared_informer.go#L114
+	// If the pod does not exist: add it
+	// If the pod exists, and the UIDs don't match, this indicates a pod
+	// delete/add that was merged into an update, so reflect the update in dsw.
+	scheduledPod, podExists := volumeObj.scheduledPods[podName]
```

Comment:
I think podName here is podUID, not the actual pod name.
https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/volume/attachdetach/util/util.go#L258
So we use podUID in desired and actual state. when pod add or update event, no matter what pod name is, pod will be added if poduid is different. 

For the missing pod deletion event, desired_state_of_world_populator will check pod with poduid, if it does not match, it will delete from desired state. https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/volume/attachdetach/populator/desired_state_of_world_populator.go#L145

Reply:
Ah, you are right. I was interpreting "podName" as the name of the pod, not the UID. So even if we do miss a deletion event for Pod with UID U1 (and only get an update event for new Pod with UID U2), desired_state_of_world_populator will eventually delete the stale pod entry for U1 eventually. And we'll have U1 and U2 in scheduledPods map for a period of time, which shouldn't cause any issues.

Hmm, ok. This is not the source of the bug I was seeing. I will need to re-investigate.